File: ies/music-topos/src/music_topos/curriculum.clj
(ns music-topos.curriculum
  "Pure curriculum compiler - generates ScoreEvents without side effects.
   Compiles curriculum sections into sequences of ScoreEvents for playback."
  (:require [music-topos.score :as score]
            [music-topos.core :as core]))

;; ============================================================================
;; PURE EVENT GENERATORS
;; ============================================================================

(defn pitch-class-events
  "Generate events for a sequence of pitches.
   Returns seq of ScoreEvents starting at start-beat."
  [start-beat pitches dur gap]
  (map-indexed
   (fn [i pitch]
     (score/make-score-event
      {:id (keyword (str "pitch-" i))
       :at (+ start-beat (* i (+ dur gap)))
       :dur dur
       :world-object (score/make-world-object
                      {:type :pitch-space
                       :value {:midi pitch :index i}})
       :audio (score/audio-from-midi pitch :amplitude 0.2)}))
   pitches))

(defn harmonic-function-events
  "Generate events for harmonic functions (T, S, D, etc.).
   Each function contains a chord."
  [start-beat functions dur gap]
  (map-indexed
   (fn [i hf]
     (score/make-score-event
      {:id (keyword (str "harmonic-" (:name hf)))
       :at (+ start-beat (* i (+ dur gap)))
       :dur dur
       :world-object (score/make-world-object
                      {:type :harmonic-world
                       :value {:name (:name hf)
                               :root (:root hf)
                               :chord (:chord hf)}})
       :audio (score/audio-from-midi (:chord hf) :amplitude 0.2)}))
   functions))

(defn modulation-events
  "Generate events for modulation intervals (circle of fifths, etc.)."
  [start-beat intervals base-pitch dur gap]
  (map-indexed
   (fn [i interval]
     (let [pitch (+ base-pitch interval)]
       (score/make-score-event
        {:id (keyword (str "modulation-" i))
         :at (+ start-beat (* i (+ dur gap)))
         :dur dur
         :world-object (score/make-world-object
                        {:type :transformation-world
                         :value {:interval interval
                                 :pitch pitch
                                 :index i}
                         :morphisms [:transpose]})
         :audio (score/audio-from-midi pitch :amplitude 0.25)})))
   intervals))

(defn chord-progression-events
  "Generate events for a chord progression."
  [start-beat chords dur gap]
  (map-indexed
   (fn [i chord]
     (let [notes (if (map? chord) (:notes chord) chord)
           name (if (map? chord) (:name chord) (str "chord-" i))]
       (score/make-score-event
        {:id (keyword (str "chord-" i))
         :at (+ start-beat (* i (+ dur gap)))
         :dur dur
         :world-object (score/make-world-object
                        {:type :chord-space
                         :value {:name name :notes notes}})
         :audio (score/audio-from-midi notes :amplitude 0.2)})))
   chords))

(defn single-pitch-event
  "Generate a single pitch event."
  [id at dur pitch amplitude]
  (score/make-score-event
   {:id id
    :at at
    :dur dur
    :world-object (score/make-world-object
                   {:type :pitch-space
                    :value {:midi pitch}})
    :audio (score/audio-from-midi pitch :amplitude amplitude)}))

;; ============================================================================
;; INITIAL WORLD COMPILER
;; ============================================================================

(defn compile-initial-world
  "Compile the initial-world into a sequence of ScoreEvents.
   Returns pure data - a vector of ScoreEvents.
   
   Structure:
   - Section 1: Initial pitch C4 at beat 0
   - Section 2: Pitch classes at beat 3  
   - Section 3: Harmonic functions at beat 5
   - Section 4: Modulation (circle of fifths) at beat 10"
  [tempo]
  (let [world core/initial-world]
    (vec
     (concat
      ;; Section 1: Initial pitch (C4)
      [(single-pitch-event :initial-pitch 0 2.0 60 0.3)]
      
      ;; Section 2: All 12 pitch classes
      (pitch-class-events 3 (:pitch-classes world) 0.1 0.05)
      
      ;; Section 3: Harmonic functions (T→S→D→T)
      (harmonic-function-events 5 (:harmonic-functions world) 0.8 0.4)
      
      ;; Section 4: Modulation (circle of fifths)
      (modulation-events 10 (:modulation-keys world) 60 0.5 0.2)))))

;; ============================================================================
;; TERMINAL WORLD COMPILER (Sonata Form)
;; ============================================================================

(defn compile-terminal-world
  "Compile the terminal-world (sonata form) into ScoreEvents.
   
   Structure:
   - Exposition: Theme 1 and Theme 2
   - Development: Key exploration
   - Recapitulation: Theme 1 return
   - Coda: Cadence and final chord"
  [tempo]
  (let [world core/terminal-world
        structure (:structure world)
        exposition (:exposition structure)
        development (:development structure)
        coda (:coda structure)]
    (vec
     (concat
      ;; EXPOSITION: Theme 1 at beat 0, Theme 2 at beat 2
      [(score/make-score-event
        {:id :exposition-theme-1
         :at 0
         :dur 0.8
         :world-object (score/make-world-object
                        {:type :chord-space
                         :value {:section :exposition
                                 :theme 1
                                 :notes (first (:themes exposition))}})
         :audio (score/audio-from-midi (first (:themes exposition)) :amplitude 0.2)})
       
       (score/make-score-event
        {:id :exposition-theme-2
         :at 2
         :dur 0.8
         :world-object (score/make-world-object
                        {:type :chord-space
                         :value {:section :exposition
                                 :theme 2
                                 :notes (second (:themes exposition))}})
         :audio (score/audio-from-midi (second (:themes exposition)) :amplitude 0.2)})]
      
      ;; DEVELOPMENT: Key exploration starting at beat 4
      (map-indexed
       (fn [i pitch]
         (score/make-score-event
          {:id (keyword (str "development-" i))
           :at (+ 4 (* i 0.6))
           :dur 0.4
           :world-object (score/make-world-object
                          {:type :pitch-space
                           :value {:section :development
                                   :exploration true
                                   :midi pitch}})
           :audio (score/audio-from-midi pitch :amplitude 0.2)}))
       (:keys development))
      
      ;; RECAPITULATION: Theme 1 return at beat 8
      [(score/make-score-event
        {:id :recapitulation-theme-1
         :at 8
         :dur 0.8
         :world-object (score/make-world-object
                        {:type :chord-space
                         :value {:section :recapitulation
                                 :return true
                                 :notes (first (:themes exposition))}})
         :audio (score/audio-from-midi (first (:themes exposition)) :amplitude 0.2)})]
      
      ;; CODA: Cadence notes at beat 10, final chord at beat 11
      (map-indexed
       (fn [i note]
         (score/make-score-event
          {:id (keyword (str "coda-cadence-" i))
           :at (+ 10 (* i 0.5))
           :dur 0.4
           :world-object (score/make-world-object
                          {:type :pitch-space
                           :value {:section :coda
                                   :cadence true
                                   :midi note}})
           :audio (score/audio-from-midi note :amplitude 0.3)}))
       (:cadence coda))
      
      ;; Final chord
      [(score/make-score-event
        {:id :coda-final
         :at 12
         :dur 2.0
         :world-object (score/make-world-object
                        {:type :chord-space
                         :value {:section :coda
                                 :final true
                                 :notes (first (:themes exposition))}})
         :audio (score/audio-from-midi (first (:themes exposition)) :amplitude 0.3)})]))))

;; ============================================================================
;; FULL CURRICULUM COMPILER
;; ============================================================================

(defn compile-full-curriculum
  "Compile all curriculum worlds into a single score.
   Returns {:tempo tempo :events [...]} with all ScoreEvents."
  ([]
   (compile-full-curriculum 120))
  ([tempo]
   (let [initial-events (compile-initial-world tempo)
         ;; Terminal world starts after initial world ends
         ;; Calculate offset based on last event in initial world
         initial-end (if (seq initial-events)
                       (let [last-evt (last initial-events)]
                         (+ (:at last-evt) (:dur last-evt)))
                       0)
         terminal-offset (+ initial-end 2) ;; 2 beat gap
         terminal-events (compile-terminal-world tempo)
         ;; Offset terminal events
         offset-terminal (mapv
                          (fn [evt]
                            (assoc evt :at (+ (:at evt) terminal-offset)))
                          terminal-events)]
     {:tempo tempo
      :worlds [{:name "initial-world"
                :start-beat 0
                :events initial-events}
               {:name "terminal-world"
                :start-beat terminal-offset
                :events offset-terminal}]
      :events (vec (concat initial-events offset-terminal))})))

;; ============================================================================
;; UTILITY FUNCTIONS
;; ============================================================================

(defn score-duration
  "Calculate total duration of a score in beats."
  [score]
  (if-let [events (seq (:events score))]
    (let [last-evt (apply max-key #(+ (:at %) (:dur %)) events)]
      (+ (:at last-evt) (:dur last-evt)))
    0))

(defn score-duration-seconds
  "Calculate total duration of a score in seconds."
  [score]
  (score/beats->seconds (:tempo score) (score-duration score)))

(defn events-at-beat
  "Get all events that are sounding at a given beat."
  [events beat]
  (filter
   (fn [evt]
     (and (<= (:at evt) beat)
          (> (+ (:at evt) (:dur evt)) beat)))
   events))

(comment
  ;; Development/testing
  (def test-score (compile-full-curriculum 120))
  (count (:events test-score))
  (score-duration test-score)
  (score-duration-seconds test-score)
  
  ;; Inspect initial world events
  (compile-initial-world 120)
  
  ;; Inspect terminal world events  
  (compile-terminal-world 120))

File: ies/music-topos/lib/curriculum.rb
# Curriculum compiler for music-topos
# Pure functions that compile curriculum sections into ScoreEvent arrays
# Mirrors src/music_topos/core.clj initial-world and terminal-world

require_relative 'score_event'

module MusicTopos
  module Curriculum
    INITIAL_WORLD = {
      name: "Initial Object World",
      pitch_classes: (60..71).to_a,
      harmonic_functions: [
        { name: "Tonic", root: 60, chord: [60, 64, 67] },
        { name: "Subdominant", root: 65, chord: [65, 69, 72] },
        { name: "Dominant", root: 67, chord: [67, 71, 74] }
      ],
      modulation_keys: [0, 7, 2, 5],
      chords: [
        { name: "C Major", notes: [60, 64, 67] },
        { name: "F Major", notes: [65, 69, 72] },
        { name: "G Major", notes: [67, 71, 74] }
      ],
      progression: [60, 65, 67, 60],
      cadences: [
        { name: "Authentic", notes: [67, 60] },
        { name: "Plagal", notes: [65, 60] },
        { name: "Half", notes: [60, 67] },
        { name: "Deceptive", notes: [67, 69] }
      ]
    }.freeze

    TERMINAL_WORLD = {
      name: "Terminal Object World (Sonata Form)",
      structure: {
        exposition: { key: 60, themes: [[60, 64, 67], [67, 71, 74]] },
        development: { keys: [65, 69, 72, 74, 78, 81], exploration: true },
        recapitulation: { key: 60, return: true },
        coda: { cadence: [67, 60], duration: 2 }
      },
      fragments: {
        pitch_classes: (60..71).to_a,
        harmonic_functions: [
          { name: "Tonic", embedding: :all_sections },
          { name: "Subdominant", embedding: :development },
          { name: "Dominant", embedding: :all_sections }
        ],
        modulation_keys: [60, 67, 74, 65, 60],
        chords: [[60, 64, 67], [65, 69, 72], [67, 71, 74]],
        progression: [[60, 64, 67], [65, 69, 72], [67, 71, 74], [60, 64, 67]],
        cadences: [[67, 74, 60], [65, 69, 60], [60, 67], [67, 69]]
      }
    }.freeze

    class << self
      def pitch_class_events(start_beat, pitches, dur, gap)
        pitches.each_with_index.map do |pitch, idx|
          ScoreEvent.new(
            id: "pitch_class_#{idx}",
            at: start_beat + (idx * (dur + gap)),
            dur: dur,
            world_object: WorldObject.new(type: :pitch_space, value: pitch),
            audio: Audio.from_midi(pitch, amplitude: 0.2),
            meta: { section: :pitch_classes, index: idx }
          )
        end
      end

      def harmonic_function_events(start_beat, functions, dur, gap)
        functions.each_with_index.map do |func, idx|
          chord = func[:chord]
          ScoreEvent.new(
            id: "harmonic_#{func[:name].downcase}",
            at: start_beat + (idx * (dur + gap)),
            dur: dur,
            world_object: WorldObject.new(type: :chord_space, value: chord),
            audio: Audio.from_midi(chord, amplitude: 0.2),
            meta: { section: :harmonic_functions, name: func[:name], index: idx }
          )
        end
      end

      def modulation_events(start_beat, intervals, dur, gap)
        intervals.each_with_index.map do |interval, idx|
          pitch = 60 + interval
          ScoreEvent.new(
            id: "modulation_#{idx}",
            at: start_beat + (idx * (dur + gap)),
            dur: dur,
            world_object: WorldObject.new(type: :transformation_world, value: interval),
            audio: Audio.from_midi(pitch, amplitude: 0.25),
            meta: { section: :modulation, interval: interval, index: idx }
          )
        end
      end

      def chord_progression_events(start_beat, chords, dur, gap)
        chords.each_with_index.map do |chord, idx|
          notes = chord.is_a?(Hash) ? chord[:notes] : chord
          ScoreEvent.new(
            id: "chord_#{idx}",
            at: start_beat + (idx * (dur + gap)),
            dur: dur,
            world_object: WorldObject.new(type: :chord_space, value: notes),
            audio: Audio.from_midi(notes, amplitude: 0.2),
            meta: { section: :chord_progression, index: idx }
          )
        end
      end

      def compile_initial_world
        events = []
        beat = 0.0

        events.concat(pitch_class_events(beat, INITIAL_WORLD[:pitch_classes], 0.08, 0.02))
        beat += INITIAL_WORLD[:pitch_classes].size * 0.1 + 0.5

        events.concat(harmonic_function_events(beat, INITIAL_WORLD[:harmonic_functions], 0.8, 0.4))
        beat += INITIAL_WORLD[:harmonic_functions].size * 1.2 + 0.5

        events.concat(modulation_events(beat, INITIAL_WORLD[:modulation_keys], 0.5, 0.1))

        events
      end

      def compile_terminal_world
        events = []
        beat = 0.0
        structure = TERMINAL_WORLD[:structure]

        events.concat(chord_progression_events(beat, structure[:exposition][:themes], 0.8, 0.2))
        beat += structure[:exposition][:themes].size * 1.0 + 0.5

        events.concat(chord_progression_events(beat, structure[:development][:keys].map { |k| [k] }, 0.3, 0.1))
        beat += structure[:development][:keys].size * 0.4 + 0.5

        events.concat(chord_progression_events(beat, [[structure[:recapitulation][:key], structure[:recapitulation][:key] + 4, structure[:recapitulation][:key] + 7]], 1.0, 0.2))
        beat += 1.2

        cadence = structure[:coda][:cadence]
        events.concat(chord_progression_events(beat, [[cadence[0]], [cadence[1]]], 1.0, 0.5))

        events
      end

      def compile_full_curriculum
        initial_events = compile_initial_world
        initial_duration = initial_events.map { |e| e.at + e.dur }.max + 1.0

        terminal_events = compile_terminal_world.map do |e|
          ScoreEvent.new(
            id: e.id,
            at: e.at + initial_duration,
            dur: e.dur,
            world_object: e.world_object,
            audio: e.audio,
            meta: e.meta
          )
        end

        initial_events + terminal_events
      end
    end
  end
end
